Prototype implementation

kelloggm opened the pull request:
This is a prototype implementation of a checker similar to the one @mernst and I discussed last week for tracing literals through source code and enforcing that they aren't passed to certain libraries.

There are a few differences from the design we discussed:
* this prototype doesn't use the Value Checker - instead, it uses `@ImplicitFor(LiteralKind.ALL)`. The reason is that using the Value Checker caused the left-hand sides of pseudo-assignments that the Value Checker had propagated to to be defaulted to `@MaybeConstant`, even if the user wrote `@NonConstant`, causing false negatives. If I can figure out a way around that, this checker could go back to defaulting based on the Value Checker, but this is sufficient for now.
* This checker forces the `-AuseDefaultsForUncheckedCode=bytecode` option. Otherwise, passing a constant to a library (like to `Random#nextInt`, which shows up in the tests) issues an error because the defaulting rules are not applied. Libraries that need to be protected must do so with stub files, which is a weakness of this design.
* A large number of features were stolen from awslabs/data-classification-checker to make the checker behave as expected on unannotated code. They all appear in the visitor.

kelloggm committed:
basic basis for the typechecker

kelloggm committed:
working prototype

kelloggm committed:
add tests

kelloggm committed:
add documentation to the qualifiers

kelloggm committed:
move and update docs into the README

kelloggm committed:
add test cases and documentation

kelloggm committed:
remove problem with defaulting for new instances, add a corresponding test, and add some more documentation to some other tests

kelloggm committed:
version of prototype that works on 2.9.0 and succeeds at correctly typechecking a simple example with WPI

kelloggm commented:
@mernst, this prototype is now at the stage where I think it should be reviewed. It passes all of its tests, and I've tested it on a simple example using the whole program inference tool in the Checker Framework; it produced the correct result.

I'm going to move on in the meantime to trying this on a larger example for now, but I'd like to get this basic version merged.

kelloggm committed:
change defaulting rules for null and for booleans so that their literals are non-constant, with tests

kelloggm committed:
fix up generics

kelloggm committed:
permit overrides of methods defined in bytecode

kelloggm committed:
add another test case

kelloggm committed:
remove now-unneeded override of isValidUse, thanks to Suzanne's improvements to defaulting in 2.9.0

mernst commented on README.md:
```diff
@@ -1,38 +1,64 @@
+
+### How does it work?
+
+The checker is built on the [Checker Framework](checkerframework.org).
```
Please start with information for users, such as what the qualifiers are and what they mean.
Implementation details are not as useful for users, and may be offputting.

kelloggm committed:
upgrade to latest versions of CF, gradle

mernst commented on README.md:
```diff
@@ -1,38 +1,64 @@
+only the bytecode is available, such as code from a `.jar` file). The following
+optimistic defaulting rules are applied:
+* the return type of any function defined in a library is `@NonConstant`
+* the types of a libraries' formal parameters are always `@MaybeConstant` unless
```
You tend to use odd rules for possessives.

To form a possessive:
 * Use 's if the noun is singular or is a plural that does not end in "s".
 * Use ' if the noun is plural *and* ends in "s".

This one should be "library's".

mernst commented on README.md:
```diff
@@ -1,38 +1,64 @@
+import org.checkerframework.checker.noliteral.qual.NonConstant;
+
+class SecretKeySpec {
+    SecretKeySpec(@NonConstant byte [] key, String algorithm);
```
Does the array itself need to be non-constant?  This is worth discussing.  The examples above did show arrays that were treated as non-constant.

kelloggm replied:
It does not.

More generally, I don't think it makes sense to think of any array as "maybe constant", nor to think of any other data structure that way. I don't think the `@MaybeConstant` type should ever be applied to any type that's not one of the manifest literals we're tracking.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/qual/MaybeConstant.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * The top type for the NoLiteral Checker's type system.
+ *
+ * <p>This type represents an expression that might contain a literal, like 5 or {0xa, 0xb}. It is
```
"expression contains" is odd phrasing, since it sounds local and syntactic.  I think you mean something more like it could be derived/computed (in part) from a literal.  In this case, `@MaybeConstant` is a somewhat confusing name for the concept, since `"hello" + readInput()` is definitely not constant (from a user perspective) but has type `@MaybeConstant`.

kelloggm replied:
I've renamed it to `MaybeDerivedFromConstant`, which I think is much clearer. Thanks for pointing this out.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/qual/MaybeConstant.java:
```diff
@@ -0,0 +1,36 @@
+ * The top type for the NoLiteral Checker's type system.
+ *
+ * <p>This type represents an expression that might contain a literal, like 5 or {0xa, 0xb}. It is
+ * inferred for all manifest literals in the program.
```
"It is inferred for" is an implementation detail.  Just say those literals have the type, or omit it since it is obvious from the first sentence.

mernst commented on README.md:
````diff
@@ -1,38 +1,64 @@
+```
+
+The default is `@NonConstant` for all expressions except manifest literals (examples
+of manifest literals include `1`, `"hello"`, and `{0xa, 0xb}`). That default includes
````
What about `{0xa, myVariable}`?

kelloggm replied:
`0xa` is a manifest literal, so the type of the array is `@MaybeDerivedFromConstant byte []`

mernst commented on src/main/java/org/checkerframework/checker/noliteral/qual/MaybeConstant.java:
```diff
@@ -0,0 +1,36 @@
+ * <p>This type represents an expression that might contain a literal, like 5 or {0xa, 0xb}. It is
+ * inferred for all manifest literals in the program.
+ *
+ * <p>This is also the type of any expression that contains a subexpression that may be a constant.
```
The Javadoc for one of the annotations is an unexpected place to find the expression typing rules.  They belong in an overview document, not here.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/qual/NonConstant.java:
```diff
@@ -0,0 +1,29 @@
+import org.checkerframework.framework.qual.TypeUseLocation;
+
+/**
+ * The bottom type for the NoLiteral Checker's type system.
```
This documentation never says what `@NonConstant` means, which is the most important thing it should say.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,42 @@
+
+/**
+ * The type factory for the no literal checker. It doesn't do much: all it provides are some
+ * canonical annotations. If we want to change this checker so that it uses the Value Checker for
```
Say why.  For example, would that reduce false negatives (improve soundness) or false positives (improve precision)?
I think the point is to identify more possible constant values than just literals in the source code.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,42 @@
+    this.postInit();
+  }
+
+  public AnnotationMirror getCanonicalBottomAnnotation() {
```
I would remove "Canonical" from the name.
Also, please write Javadoc documentation.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralChecker.java:
```diff
@@ -0,0 +1,22 @@
+
+  /**
+   * Overridden so that unchecked code defaults are always used for bytecode, because we only want
+   * to enforce @NonConstant if a stub file is present.
```
I think the point of this is to prevent defaulting to `@NonConstant`.  If so, please say that.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+import org.checkerframework.javacutil.ElementUtils;
+
+/**
+ * This visitor has a lot of code in it that permits this type system to operate as expected with
```
Should you create a new class `BottomDefaultTypeVisitor`, and have both systems use it?
That would reduce code, avoid cut-and-paste errors, and ensure that fixes propagate to all implementations.

kelloggm replied:
The Checker Framework probably should provide such a class. Let's plan on writing one and getting it reviewed after the FSE deadline.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+
+/**
+ * This visitor has a lot of code in it that permits this type system to operate as expected with
+ * the bottom type as its default. Most of that code is either from or inspired by AWS' Data
```
Please fix pluralization.

kelloggm replied:
I rewrote this so that the possessive could be removed, but this is actually an interesting grammar corner case. AWS is an initialism, so you would normally using 's when turning it into its possessive form. However, the S in AWS stands for a plural noun ("Services"), so if you wrote it out, it would be "Amazon Web Services'". I'm actually not sure what's correct: a (very good use of time) 10 minute google search turned up multiple results that give each answer to similar questions, though I think there's a bit of a bias toward "AWS's" rather than "AWS'", it's certainly not settled.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+  }
+
+  /**
+   * Overrides the default lower bound for exception parameters. By default, this is top. This
```
I don't know whether the "This" at the end of the line (the start of the third sentence) refers to the first sentence or the second.  Please clarify the writing.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+  /**
+   * Overrides the default lower bound for exception parameters. By default, this is top. This
+   * prevents the checker from defaulting exception parameters to @NonConstant. Overriding this
+   * method changes the expected lower bound to be @NonConstant, permitting that default.
```
What does "expected" mean in this context?

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+            (AnnotatedTypeMirror.AnnotatedDeclaredType) atm;
+        for (AnnotatedTypeMirror component : atmD.getTypeArguments()) {
+          if (hasNonDefault(component)) {
+            result = true;
```
Why isn't this `return true;`?

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+        AnnotatedTypeMirror.AnnotatedArrayType atmA = (AnnotatedTypeMirror.AnnotatedArrayType) atm;
+        result = hasNonDefault(atmA.getComponentType());
+        break;
+      case WILDCARD:
```
Might there be non-default types on its bound?  Or is that taken care of by the base case?

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+        return false;
+      case TYPEVAR:
+        return false;
+      default:
```
A default case that doesn't throw an exception is a code smell.  Can you expand this to show the cases for which the switch statement does nothing?

More generally, it might be nicer to write a TypeVisitor to do this work, rather than the current code structure.  That will avoid various potential errors and be more future-proof.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+      default:
+        break;
+    }
+    return result || !atm.hasAnnotation(NonConstant.class);
```
Do the base case first.  It is more efficient to check it first rather than after iterating through the type parameters and array levels.  I think it is also less confusing.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+  /**
+   * This checker defaults implicit upper bounds of type variables to @NonConstant. This is the
+   * correct thing to do in unannotated code, but causes false positive errors whenever a user
+   * writes a type annotation on the upper bound of a type variable (for instance, by declaring a
```
Please give a concrete example rather than just "by declaring a list of constant strings".

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+   * This checker defaults implicit upper bounds of type variables to @NonConstant. This is the
+   * correct thing to do in unannotated code, but causes false positive errors whenever a user
+   * writes a type annotation on the upper bound of a type variable (for instance, by declaring a
+   * list of constant strings). Since this is common, this code disables that check for user-written
```
The documentation doesn't say what "that check" is.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+  }
+
+  /**
+   * Skip the standard subtyping check on method calls always. Constants can't have methods called
```
"Skip the standard subtyping check on method calls" is misleading, since only part of that check (the receiver part) is skipped.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+
+  /**
+   * Skip the standard subtyping check on method calls always. Constants can't have methods called
+   * on them, so we don't care.
```
Some constants, such as strings, can have methods called on them.  I would expect String methods to be polymorphic in their arguments.  I don't see any test cases for string method calls.

I was surprised not to find a polymorphic annotation in this type system, since it is useful for most type systems.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+      AnnotatedExecutableType constructorType, ExecutableElement constructorElement) {
+    // Skip this and do nothing. Normally, this method will issue a warning when
+    // the type of a class defaults to something other than top. That's fine
+    // for this checker, because we don't expect user-written classes to take
```
The checker could verify this assumption.

kelloggm replied:
It could. I will open an issue.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    @Override
+    public boolean checkOverride() {
```
Please say what this is doing (that's different than the superclass).  That might be a code comment at the beginning of the method rather than a Javadoc comment -- I don't know what is appropriate.

kelloggm replied:
The documentation appears in the Javadoc comment for `NoLiteralVisitor#createOverrideChecker()`, which I thought (and still think) is the most appropriate place for it. I will add a summary comment here, however.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/NoLiteralChecker.java:
```diff
@@ -0,0 +1,22 @@
+public class NoLiteralChecker extends BaseTypeChecker {
+
+  /**
+   * Overridden so that unchecked code defaults are always used for bytecode, because we only want
```
Give the specification for the method.  Say what it does rather than starting off with an obvious fact about the implementation (that it is overridden),  When you write something, readers will assume it is important, and when it's the first thing you write, readers should assume it is the most important thing.  If that is not the case, the writing is confusing.

mernst commented on src/main/java/org/checkerframework/checker/noliteral/qual/MaybeConstant.java:
```diff
@@ -0,0 +1,36 @@
+import org.checkerframework.framework.qual.TypeUseLocation;
+
+/**
+ * The top type for the NoLiteral Checker's type system.
```
Start by saying what it means.  This fact is obvious from the annotations and is not the most important information that a reader of the Javadoc will want to know.

mernst commented on tests/noliteral/FloatInference.java:
```diff
@@ -0,0 +1,10 @@
+import org.checkerframework.checker.noliteral.qual.*;
```
I found the profusion of tiny test files confusing.  It's better to organize the tests.  For example, put all the `*Inference` tests in a single file since they are testing essentially the same thing.  Then it's easier to look over the list of tests and understand what they are testing and what they are not.

mernst reviewed (changes requested):
I have two general comments.

**First**, the specification for this checker is not clear.
I think that the rule is:
 * any manifest constant has type @MaybeLiteral
 * any computation derived from a @MaybeLiteral has type @MaybeLiteral
In other words, @MaybeLiteral is the forward slice from manifest constants, or the set of computations reachable from a manifest constant.

Stating a rule like this would clarify the type system for both users and implementers.

**Second**, when you write, try to keep in mind the reader and focus on their concerns.  This is really hard to do!  In a paper, you might mention only briefly the things that required the greatest amount of time.  In a specification, you will gloss over implementation details that took a lot of effort.

You might think it a disservice to omit true details.  But including them can dissuade the reader from continuing to read or can distract the reader from the most important issues.  Put less important details at the end or in auxiliary documents.

**Finally**, a minor comment:  the date in the `LICENSE` file should be updated in this pull request.

kelloggm committed:
make spec more clear, and rename an annotation to match

kelloggm committed:
change license date

kelloggm committed:
fix typo + aesthetic changes in README

kelloggm committed:
fix typo

kelloggm committed:
add more text on how to use the checkeR

kelloggm committed:
rename some methods and improve a comment

kelloggm committed:
remove unhelpful comments

kelloggm committed:
rename some methods again, add Javadoc

kelloggm committed:
remove unnecessary annotation

kelloggm committed:
rewrite comment

kelloggm committed:
grammar fix

kelloggm committed:
formatting, and improve a comment

kelloggm committed:
simplify logic

kelloggm committed:
replace a bad method with a better visitor

kelloggm committed:
improve javadoc on two methods

kelloggm committed:
fix formatting, improve javadoc more

kelloggm committed:
add another comment saying the same thing

kelloggm committed:
consolidate some tests

kelloggm committed:
add polymorphic qualifier

kelloggm committed:
add stub for java.lang.integer

kelloggm committed:
add java.lang.Byte stub

kelloggm committed:
add String stub file + tests

kelloggm committed:
add print class path task

kelloggm committed:
redo how the repo is setup so that qual and checker are separate

kelloggm committed:
small extra test

kelloggm committed:
fix bug with external libraries

kelloggm committed:
toString, hashCode, clone specs

kelloggm committed:
fix defaulting for array component types in bytecode

kelloggm committed:
handle generics better

kelloggm committed:
formatting

kelloggm committed:
consistency with other defaulting for method return types

kelloggm committed:
fix defaulting rules for type variables by using visitDeclared instead of a collection of workarounds

kelloggm committed:
add models for Short and Long

kelloggm committed:
formatting

kelloggm committed:
commit test case that shows a bit of weird behavior wrt where errors are issued, but also demos that there's no fp on correct codE

kelloggm committed:
add warning about prototype status

kelloggm committed:
remove comment

kelloggm committed:
explain a weird line of code

kelloggm committed:
specify what the test does

kelloggm committed:
add newline

kelloggm committed:
add newlines to tests

kelloggm committed:
make sure all tests have a note explaining what they're testing

kelloggm commented:
@mernst, this is ready for review again. It's still clearly a prototype, so I'm mostly interested in your opinions on the design decisions and the reasoning behind them, or any test cases that you think I've missed.

I've run it on a single 2300-ish line project, and it only issues the 4 expected errors (though two of them in the wrong place - see the test `SecretKeySpecTestUnnannotated.java`, which models those errors).

kelloggm committed:
avoid annotation false positives

kelloggm committed:
test case and some exploratory work that doesn't do the right thing

kelloggm committed:
fix some annotations on String

kelloggm committed:
float and double stubs + tests

mernst commented on README.md:
```diff
@@ -1,38 +1,75 @@
+
+### How does it work?
+
+The checker computes a forward slice from each manifest literal in the program
```
It doesn't actually compute a forward slice.
Also, "forward slice" is jargon that users are unlikely to be familiar with.
Please reword.

mernst commented on README.md:
```diff
@@ -1,38 +1,75 @@
+optimistic defaulting rules are applied:
+* the return type of any function defined in a library is `@NonConstant`
+* the types of a library function's formal parameters are always `@MaybeDerivedFromConstant` 
+unless a stub file is supplied that overwrites this default
```
Does this relate to just the last bullet point or to both bullet points?
If the former, indent by two spaces.
If the latter, add a blank line before it.

mernst commented on no-literal-qual/src/main/java/org/checkerframework/checker/noliteral/qual/PolyConstant.java:
```diff
@@ -0,0 +1,15 @@
+import java.lang.annotation.Target;
+import org.checkerframework.framework.qual.PolymorphicQualifier;
+
+/** A polymorphic qualifier for the no literal type system. */
```
Each qualifier should link back to the manual (polymorphic ones should link to both the checker's manual and the Checker Framework Manual discussion of polymorphism), for context for users viewing the formatted API documentation.

kelloggm replied:
This checker doesn't have a manual section (because it's not in the Checker Framework's source tree), so I clearly cannot link to it.

I will add a link to the section about polymorphic qualifiers.

mernst replied:
It does have a manual at https://github.com/kelloggm/no-literal-checker, and it is worth linking to the manual.

kelloggm replied:
Ah I misunderstood your comment to be asking for a link to the Checker Framework manual specifically. I've added links to the `README.md` file.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  @Override
+  protected TypeAnnotator createTypeAnnotator() {
```
Is this method definition needed?  If so, that seems like a bug in the Checker Framework, because it should not require writing boilerplate implementations.

kelloggm replied:
Yes, it is needed, both for subclasses of `TypeAnnotator` and of `TreeAnnotator`. See for example [the Interning Checker](https://github.com/typetools/checker-framework/blob/master/checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java#L124) or [the Lower Bound Checker](https://github.com/typetools/checker-framework/blob/master/checker/src/main/java/org/checkerframework/checker/index/lowerbound/LowerBoundAnnotatedTypeFactory.java#L233). It is a pain that these are needed, but it's been true for a long time and was never an issue before. I certainly don't object to you opening a bug report, but I have no interest in shepherding one or fixing it, because I don't think it's that inconvenient.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,191 @@
+    QualifierDefaults defaults = new NoLiteralQualifierDefaults(elements, this);
+    // Ensure that unchecked code is treated optimistically by switching the
+    // standard defaults for unchecked code.
+    defaults.addUncheckedCodeDefault(MAYBE_CONSTANT, TypeUseLocation.PARAMETER);
```
The comment on lines 59-60 seems to refer to the code on lines 62-64 but not 61.  Please clarify or move the comment.

kelloggm replied:
They refer to all four lines: the standard default at `TypeUseLocation.PARAMETER` would be `@NonConstant`, and at the other locations on lines 62-64 would be `@MaybeConstant`. So for both cases this method is switching the defaults.

I'll add new lines around this to make it clear.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,191 @@
+    // Ensure that unchecked code is treated optimistically by switching the
+    // standard defaults for unchecked code.
+    defaults.addUncheckedCodeDefault(MAYBE_CONSTANT, TypeUseLocation.PARAMETER);
+    defaults.addUncheckedCodeDefault(NON_CONSTANT, TypeUseLocation.RETURN);
```
Please don't hardcode this list.  Instead, use `QualifierDefault.STANDARD_UNCHECKED_DEFAULTS_TOP`:  either make it public, or add a method similar to `addUncheckedStandardDefaults` that you can call from here and from any other checker that wishes for similar behavior.

kelloggm replied:
I can (and will) use `STANDARD_UNCHECKED_DEFAULTS_TOP`, which I did not notice before. However, I cannot use `STANDARD_UNCHECKED_DEFAULTS_BOTTOM`, because I don't want to change the default for lower bounds, only for parameters (because of the other changes I make to defaulting for type variables later). I'll add a comment.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,191 @@
+  /**
+   * Returns true if the given type mirror represents a declared type that could have been derived
+   * from a literal: a boxed primitive or a string. Returns false for java.lang.Boolean, because
+   * this checker does not consider booleans literals.
```
That fact is worth stating in the manual.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,259 @@
+
+/**
+ * This visitor has a lot of code in it that permits this type system to operate as expected with
+ * the bottom type as its default. Most of that code is either from or inspired by the AWS Data
```
Saying something about the relationship with the Nullness Checker might be useful, since it also has the bottom type as its default.  Should there be a subtype of BaseTypeVisitor for type systems with bottom as their default?

kelloggm replied:
This checker has no direct relationship with the Nullness Checker. I did add a mention of it in the manual, though, when I mention that this checker doesn't consider `null` to be a constant.

I do think there should be a subtype of `BaseTypeVisitor` for bottom-default type systems. For right now, I'm not willing to take on that problem, though.

mernst committed:
Code review changes

mernst committed:
Delete trailing whitespace

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralChecker.java:
```diff
@@ -0,0 +1,22 @@
+
+import org.checkerframework.common.basetype.BaseTypeChecker;
+
+/** The NoLiteral Checker enforces that literals don't go where they don't belong. */
```
Be more specific.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,188 @@
+
+  @Override
+  protected void checkInvalidOptionsInferSignatures() {
+    // This checker is specifically designed to work with whole-program inference,
```
This comment is confusing, since the whole point of whole-program inference is that it works with every type system, without change.  Can you rephrase it to be more specific?

kelloggm replied:
I'm not sure how I could make this more specific. Whole program inference refuses to run with certain command line options that this checker requires (in particular, `-AuseConservativeDefaults=-bytecode`, see `NoLiteralChecker.java`). This comment says that this checker has been specifically tested with WPI and works, so the check is turned off. I don't know why the check was there in the first place, to be honest, so it could possibly be removed.

mernst commented on no-literal-qual/src/main/java/org/checkerframework/checker/noliteral/qual/MaybeDerivedFromConstant.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * This type represents an expression that may have been derived (in part) from a literal, like {@code 5}, {@code "hello"},  or {@code {0xa, 0xb}}.
+
+ * This is also the type of any expression that contains a subexpression that may be a constant. For
```
The word "also" suggests that this sentence introduces new information.  But I think this is just a consequence or an example of the rule in the first sentence.

mernst replied:
Does "that may be a constant" mean "of type `@MaybeDerivedFromConstant`"?  If so, say it that way.  If not, define it.

mernst commented on no-literal-qual/src/main/java/org/checkerframework/checker/noliteral/qual/MaybeDerivedFromConstant.java:
```diff
@@ -0,0 +1,33 @@
+import org.checkerframework.framework.qual.TypeUseLocation;
+
+/**
+ * This type represents an expression that may have been derived (in part) from a literal, like {@code 5}, {@code "hello"},  or {@code {0xa, 0xb}}.
```
Consider making "derived from" a bit more explicit.

kelloggm committed:
partial work towards a fix

kelloggm committed:
address PR feedback

kelloggm commented:
@mernst I've addressed your comments

kelloggm committed:
better links

kelloggm committed:
fix for local array component inference

kelloggm committed:
better docs

kelloggm committed:
formatting

kelloggm committed:
fix javadoc

kelloggm committed:
Merge branch 'prototype' of github.com:kelloggm/no-literal-checker into local-array-infer

kelloggm committed:
correctly handle code that's presented as both bytecode and as source code (that is, treat it as source)

kelloggm committed:
Merge pull request #7 from kelloggm/bytecode-vs-sourcecode

Correctly handle code that's presented as both bytecode and as source

kelloggm committed:
CI

kelloggm committed:
CI

kelloggm committed:
allow nlc inference logic to handle arrays that aren't NEW_ARRAY trees

kelloggm committed:
use polymorphic defaults for unchecked code

kelloggm committed:
formatting

kelloggm committed:
Merge pull request #8 from kelloggm/new-tests

Use polymorphic defaults for unchecked code

kelloggm committed:
Merge pull request #6 from kelloggm/local-array-infer

Infer component types of effectively final local arrays

kelloggm committed:
remove requirement that arrays need to be effectively final for their types to be inferred

mernst commented:
@kelloggm Could you make CI pass and resolve the conflict with master?

kelloggm committed:
Merge branch 'master' into prototype

kelloggm committed:
formatting

kelloggm committed:
working, but sketchy version. Committing so I can come back to it after some experimentation

kelloggm committed:
fill out some stub files, and add some tests

kelloggm committed:
formatting

kelloggm committed:
build on a Java 8 JVM

kelloggm committed:
correct handling of --release 8 interactions with JDK methods, and remove some unnecessary stuff that causes stub warnings from the stub files

kelloggm committed:
fix handling of return types in overrides

kelloggm committed:
remove debugging code

kelloggm committed:
handle environment variables and properties correctly

kelloggm committed:
handle javax.swing.JOptionPane showInputDialog

kelloggm committed:
better defaulting for polymorphism+defaults, plus various other minor improvements

mernst commented:
@kelloggm A ping on this pull request.

kelloggm committed:
formatting, and use CF 3.4.1

kelloggm committed:
documentation

kelloggm committed:
improve docs

kelloggm committed:
fix travis build

kelloggm commented:
@mernst, this PR is now ready for you to review again. There have been substantial changes to it since your last review; the primary ones are:
* the defaulting scheme has been changed to poly-by-default for bytecode, and
* local type inference for the component types of arrays has been implemented.

mernst commented on gradle-mvn-push.gradle:
```diff
@@ -0,0 +1,247 @@
+ */
+
+// This script is derived from the version here:
+// https://github.com/uber/android-template/blob/master/gradle/gradle-mvn-push.gradle
```
This URL gives me a 404 error.

mernst replied:
Ping on this comment.

Since this file is apparently copied from somewhere, could you add a comment at the very top of this file indicating where you copied it from?

kelloggm replied:
I removed this file - with Maven-publish this won't be necessary at all. For your information, it comes from the object construction checker repository: https://github.com/kelloggm/object-construction-checker/blob/master/gradle-mvn-push.gradle, which should also probably upgrade.

mernst commented on gradle-mvn-push.gradle:
```diff
@@ -0,0 +1,247 @@
+// This script is derived from the version here:
+// https://github.com/uber/android-template/blob/master/gradle/gradle-mvn-push.gradle
+
+apply plugin: 'maven'
```
The maven plugin is deprecated and scheduled for removal.  Use the maven-publish plugin instead.

kelloggm replied:
I removed all the references to the maven plugin and added a TODO to use maven-publish instead. We've never actually deployed this checker to Maven, so I'm going to remove publishing from this PR to speed it up; I'll re-add it later.

mernst commented on README.md:
```diff
@@ -1,38 +1,100 @@
+Any manifest literal of types `short`, `int`, `long`, `byte`, `char`, `float`,
+`double`, or `String`, or any array of these, is considered a hard-coded constant.
+Examples include `1`, `"hello"`, and `{0xa, 0xb}`. Values derived from other hard-coded
+constants by arithmetic are also considered hard-coded. For example `1 + x` is a
```
Why only arithmetic?  Should this include other operations, such as string concatenation?

kelloggm replied:
This is a typo - it should say "arithmetic-like operations, including string concatenation"

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,448 @@
+    defaults.addUncheckedCodeDefault(POLY, TypeUseLocation.RETURN);
+    defaults.addUncheckedCodeDefault(POLY, TypeUseLocation.PARAMETER);
+
+    // Fields are assumed to be non-constant. This default is unsound, but dataflow
```
Can you explain this comment?  I don't see how a default can be unsound -- all types in source code, defaulted or not, are checked by the checker.  Do you mean that it might interact unsoundly with unchecked code?  Or that the checker itself is unsound?

kelloggm replied:
This code is only handling defaulting in **unchecked** code. This default says that a field in unchecked is also assumed to be non-constant, which isn't sound (i.e. conservative) - the field may be a constant.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+    @Override
+    public boolean checkOverride() {
+      // Don't issue override errors if the overridden method was defined
+      // in bytecode, which uses a different defaulting scheme.
```
Why is the defaulting scheme relevant?  Illegal overriding is still illegal, and should be checked on types after defaulting.  It should not be cognizant of defaulting.

kelloggm replied:
If some method `a(int x)` is defined in a class `C` that is on the classpath but whose source is unavailable, and a class `D extends C` overrides `a(int x)`, then the version of `a` in `D` will be defaulted as if it is source code (that is, `x` will be `@NonConstant`, but the version of `a` in `C` will be defaulted as bytecode (i.e. `x` will be `@PolyConstant`. These types are incompatible, which is why the check is modified here.

I don't think this is a soundness concern. This code treats `C.a` as if its parameters are bottom and its return type is top - that is, treats it as if we were doing conservative defaulting. That means that the overriding method `D.a` has to follow the more restrictive conservative defaulting scheme, even though a call to `C.a` might be treated as polymorphic, which can only cause false positives (if `a` is called on an object whose static type is `C` but whose dynamic type is `D`).

At least, I think that's what's happening. It's been awhile since I wrote this code so I don't remember the original justification - this is my attempt to reconstruct it now.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  /**
+   * Skip the standard subtyping check on receivers on all method calls. Constants can't have
```
Is this skipped for efficiency or correctness?  A note would help readers understand.

kelloggm replied:
It's to suppress false positives. If for some reason we end up with an `@MaybeConstant` type that can't be a constant, we would get false positive warnings on any methods objects of that type call. I suspect I encountered that in practice, which was why this code would have been added.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+
+  /**
+   * The standard override checker is replaced so that no override.param.invalid errors are issued
+   * when overriding methods defined in bytecode (which therefore use a different, incompatible
```
I don't understand the causality in "which therefore".  Should it be "because they"?

kelloggm replied:
yes

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+
+  /**
+   * The standard override checker is replaced so that no override.param.invalid errors are issued
+   * when overriding methods defined in bytecode (which therefore use a different, incompatible
```
Why is it OK to not issue errors about illegal overriding?  Is that sound?

kelloggm replied:
See my comment above on the code that actually implements it - I think it's sound, because it basically pretends for the purpose of override checking that bytecode uses conservative defaults.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+
+    @Override
+    protected Void scan(AnnotatedTypeMirror type, Void aVoid) {
+      if (!type.hasAnnotation(NonConstant.class)) {
```
The Javadoc says that this reports about non-default annotations, but `@NonConstant` is not the default everywhere, not even everywhere in source code (e.g., type variable bounds).  Please explain or fix.

kelloggm replied:
The name was wrong; I've fixed it (it's only used in one place, so what it's actually doing is what it should be doing).

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  /**
+   * This checker defaults implicit upper bounds of type variables to @NonConstant. This is the
```
Can you give a cross-reference to where that default is implemented?

kelloggm replied:
I can't find code that implements such a default. Either that's the standard default that the Checker Framework always uses, or this comment is wrong. I suspect the comment is wrong, because the example below refers to implicit upper bounds of type variables, which are a CLIMB-to-top location.

I removed this code and re-ran the tests, and got the same results. So I'm going to remove it.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+   *
+   * <p>For example, without overriding this method, this code would not typecheck:
+   *
+   * <p>{@code List<@MaybeDerivedFromConstant String> list = new ArrayList<>();}
```
Is `new ArrayList<>()` essential to the example?  If not, then simplify this to 
```
List<@MaybeDerivedFromConstant String> list = ...;
```
or to
```
List<@MaybeDerivedFromConstant String> list;
```
in order to focus the reader's attention on what is relevant.

kelloggm replied:
This has been removed.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java:
```diff
@@ -0,0 +1,244 @@
+   *
+   * <p>{@code List<@MaybeDerivedFromConstant String> list = new ArrayList<>();}
+   *
+   * <p>Since this is common, this code disables checking that type arguments supplied to a type or
```
Disabling checking sounds scary.  Are there any possible downsides?  Is it sound?

mernst commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
```
You cannot use the MIT license, because this checker copies code from the Data Classification Checker which [uses the Apache License](https://github.com/awslabs/data-classification-checker/blob/master/LICENSE).  You either need to change this license, or you need to re-implement the relevant parts from memory in order to use the MIT license.

It will be nicer to use the MIT license, since the Apache license adds restrictions both to your code (for example, it must state chenges) and is generally more intimidating to users.

kelloggm replied:
The amount of copied code is very small. I'll re-implement it from memory.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,448 @@
+        rhsIsMaybeConstant = true; // 1 is a constant
+        break;
+      case VARIABLE:
+        return; // handled elsewhere
```
Please provide a more specific cross-reference.

mernst commented on no-literal-checker/tests/noliteral/LocalArray.java:
```diff
@@ -0,0 +1,272 @@
+import org.checkerframework.dataflow.qual.Pure;
+
+class LocalArray {
+    void test_constants() {
```
Indentation in this project is inconsistent.  The code uses 2-space indentation but the tests use 4-space indentation.  Please be consistent.  (2 spaces is more standard.)

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,448 @@
+     */
+    private class NoLiteralDefaultApplierElement extends DefaultApplierElement {
+
+      /** Whether the target type is from bytecode. */
```
Documentation is inconsistent with variable name.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,448 @@
+      }
+
+      /**
+       * Add the qualifier to the type if it does not already have an annotation in the same
```
It doesn't necessarily add the given qualifier.  It might add a different qualifier.  Could you update the documentation to reflect that?  It also doesn't mention that it applies defaults all the way down in an array.

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,448 @@
+      }
+
+      // Find the innermost component type.
+      AnnotatedTypeMirror componentType = ((AnnotatedArrayType) type).getComponentType();
```
Throughout, use `innermostComponentType` rather than `componentType` when it means the innermost one.

mernst committed:
Changes from code review

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,460 @@
+
+  /**
+   * Returns true if the given type mirror represents a type that includes values that could be
+   * derived from a literal: a boxed primitive, string, or java.lang.Object. Returns false for
```
There is also `Number`, `Serializable`, and `Comparable`.

mernst committed:
Tweak comment

mernst commented on no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,460 @@
+        if (fromSource) {
+          super.addAnnotation(type, qual);
+        } else {
+          // Use polymorphic defaulting for possibly-literal types.
```
Is the polymorphic defaulting implemented elsewhere in this checker, or implemented by virtue of what value is passed as the `qual` argument, or in some other way?

kelloggm replied:
Elsewhere in the checker. I'll add a cross-reference.

mernst commented on no-literal-checker/tests/noliteral/LocalArray.java:
```diff
@@ -0,0 +1,271 @@
+    @Pure
+    void doNothing(@MaybeDerivedFromConstant String[] arr) { }
+
+    void test_3d_array(String s) {
```
This is named test_3d_array but it seems to test only 2d arrays.

mernst commented on README.md:
```diff
@@ -1,38 +1,104 @@
+### Local type inference for array component types
+
+Unlike most typecheckers built on the Checker Framework, this checker includes custom rules
+for inferring the component types of local arrays. That is, the checker can infer that
```
Is this only the innermost component type, or any component type (which might itself be an array, as a component [in a multi-dimensional array)?

kelloggm replied:
Only the innermost component type. Array types are not considered possible constants by the checker - only the innermost component type matters. I'll update this documentation.

kelloggm replied:
Related, an empty array isn't really an exception: its "components" should be treated as the constant, not the array itself (because they're always the same - nothing!). The type of `T[] = new T[0]` should be `@MaybeDerivedFromConstant T @NonConstant []`, perhaps somewhat counter-intuitively.

mernst committed:
Commit failing tests that use parenthesization

mernst committed:
An empty array is a constant array.

kelloggm committed:
changes from PR, including fixing new failing tests involing empty arrays and parens

kelloggm commented:
@mernst I've addressed your comments. I'd really like to merge this PR before Monday evening (i.e. before ASE), so if you have any further serious concerns it's probably better to file them as issues rather than adding more comments here.

Thanks!
Martin
